Add plan query that follows the relay connection specification

optionsome opened the pull request:
### Summary

Adds a new plan query that follows the relay connection specification and some new features/enhancements compared to the old plan query in the LegacyGraphQLAPI.

Example query:
<details>
<pre>
{
  planConnection(
     dateTime: {
      earliestDeparture: "2023-06-13T14:30+03:00"
    }
    searchWindow: "PT2H30M"
    numberOfItineraries: 5
    origin: {
      location: {
        coordinate: {
          latitude: 10.0
          longitude: 20.0
        }
      }
      label: "Home"
    }
    destination: {
      location: {
        coordinate: {
          latitude: 15.0
          longitude: 25.0
        }
      }
      label: "Work"
    }
    modes: {
      directOnly: false
      transitOnly: false
      direct: [CAR_PARKING]
      transit: {
        access: [CAR_PARKING]
        transfer: [WALK]
        egress: [BICYCLE_RENTAL, WALK]
        transit: [
          {
            mode: TRAM
            cost: {
              reluctance: 1.3
            }
          },
          {
            mode: BUS
          }
        ]
      }
    }
    preferences: {
      accessibility: {
        wheelchair: {
          enabled: true
        }
      }
      street: {
        car: {
          reluctance: 6.5
          rental: {
            allowedNetworks: ["foo", "bar"]
            bannedNetworks: ["foobar"]
          }
          parking: {
            unpreferredCost: 200
            preferred: [{
              select: [{
                tags: ["best-park"]
              }]
            }]
            filters: [{
              not: [{
                tags: ["worst-park"]
              }]
            }]
          }
        }
        bicycle: {
          reluctance: 3.0
          speed: 7.4
          optimization: {
            type: SAFEST_STREETS
          }
          boardCost: 200
          walk: {
            speed: 1.3
            cost: {
              mountDismountCost: 100
              reluctance: 3.5
            }
            mountDismountTime: "PT5S"
          }
          rental: {
            destinationBicyclePolicy: {
              allowKeeping: true
              keepingCost: 300
            }
            allowedNetworks: ["foo", "bar"]
            bannedNetworks: ["foobar"]
          }
          parking: {
            unpreferredCost: 200
            preferred: [{
              select: [{
                tags: ["best-park"]
              }]
            }]
            filters: [{
              not: [{
                tags: ["worst-park"]
              }]
            }]
          }
        }
        walk: {
          speed: 2.4
          reluctance: 1.5
          walkSafetyFactor: 0.5
          boardCost: 200
        }
      }
      transit: {
        board: {
          waitReluctance: 3.2
          slack: "PT1M30S"
        }
        alight: {
          slack: "PT0S"
        }
        transfer: {
          cost: 200
          slack: "PT2M"
          maximumAdditionalTransfers: 2
          maximumTransfers: 5
        }
        timetable: {
          excludeRealTimeUpdates: false
          includePlannedCancellations: false
          includeRealTimeCancellations: true
        }
      }
    }
    locale: "en"
    after: ""
    first: 3
  ) {
    searchDateTime
    routingErrors {
      code
      description
      inputField
    }
    pageInfo {
      hasNextPage
      hasPreviousPage
      startCursor
      endCursor
    }
    edges {
      cursor
      node {
        legs {
          startTime
        }
      }
    }
  }
}
</pre>
</details>

A simple "minimal" query:

<details>
<pre>
{
  planConnection(
    dateTime: {
      earliestDeparture: "2023-06-13T14:30+03:00"
    }
    origin: {
      location: {
        coordinate: {
          latitude: 10.0
          longitude: 20.0
        }
      }
    }
    destination: {
      location: {
        coordinate: {
          latitude: 15.0
          longitude: 25.0
        }
      }
    }
    modes: {
      direct: [CAR_PARKING]
      transit: {
        access: [CAR_PARKING]
        transfer: [WALK]
        egress: [BICYCLE_RENTAL, WALK]
        transit: [
          {
            mode: TRAM
          },
          {
            mode: BUS
          }
        ]
      }
    }
  ) {
    edges {
      node {
        legs {
          startTime
        }
      }
    }
  }
}
</pre>
</details>

### Issue

closes #4803

### Unit tests

Added

### Documentation

Not needed yet

### Changelog

Will add for sandbox

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -837,6 +873,23 @@ type Coordinates {
+    arrival: SearchWindow
+}
+
+input SearchWindow {
+    dateTime: DateTime
+    timeZone: TimeZone
+    before: SearchWindowRestriction
+    after: SearchWindowRestriction
+}
+
+input SearchWindowRestriction {
+    maxOffset: Duration
+    numberOfItineraries: Int
+}
```
What should we do with paging if the search window and number of itineraries are defined like that. I guess the previous/next page logic is quite simple as we then search starting from the first/last included itinerary (depending on if it's arrive by search or not). However, deciding how many itineraries we should return for the next page is not as clear.

Also, does this model support the timetable search mode that @t2gran mentioned in https://github.com/opentripplanner/OpenTripPlanner/issues/4803#issuecomment-1429990094

t2gran replied:
The paging would be ok, but I do not think the numOf before/after is a good idea. What the `numOf..` is when paging does not matter so much, adding the two numbers would work for the next/previous window. Note! The window is scaled to fit the `numOfIt.`. Is the use-case here to allow the UI implementer to include a few options before the time the user requests to show alternatives the user actually might reach? If that is the case I suggest:
```
input SearchWindow {
    // The instance of time the search ...
    time: TimeInstance
    // The initial length of the first search window
    length: Duration
    // The requested number of itineraries. The seach window wil be adjusted to fit the number requested.
    requestedSize : Int
    // Include a few itineraries before the time if the user have a change of bording it, this will be ignored 
    // when paging back/forward    
    grecePeriod : GracePeriodInput {
       length : Duration 
       maxSize : Int
    }
}
```

optionsome replied:
Yes I think the use case is to allow returning some itineraries before the requested time. I think your suggestion might be clearer. However, I don't like the term size here as it's not really directly the size of the search window and I would expect the unit of size to be a duration.

t2gran replied:
Agree, `numberOfItineraties` is much more clear - and keep in mind I am only trying to kick you to be sure you have thought of all options - in the end you decide what to do here.

t2gran commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -758,6 +768,21 @@ enum BikesAllowed {
 }
 
+input BikePreferences {
+    reluctance: Float
+    bikeWalking: WalkPreferences
+    speed: Float
+    switchTime: Int
+    switchCost: Int
+    optimize: OptimizationCriteria
+    boardCost: Int
```
I would make Scalars for all common types we use: Speed, Duration, Cost, Slack and Reluctance

optionsome replied:
Another question here, we probably still want to have costs that only support a static cost, for example `walkBoardCost`. Then we have use cases where we want to support a cost function (or alternatively a static cost). Should we have two scalar types Cost and CostFunction? Is cost always a positive integer? Does it make sense to have input wrapper type like I have used in the schema currently (Cost should be used instead of Float):
```
input Cost @oneOf {
    static: Float
    function: CostFunction
}
```

t2gran commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1195,6 +1253,43 @@ input ParkingFilter {
 }
 
+input PlanDebugSettings {
+    debugItineraryFilter: Boolean
```
This is not a boolean flag any more, see: ItineraryFilterDebugProfile

t2gran reviewed (commented):
Some comments

t2gran commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1195,6 +1253,43 @@ input ParkingFilter {
+input PlanLocation @oneOf {
+    coordinate: Coordinate
+    """TODO should place string be a scalar?"""
+    place: String
```
I would replace place an explicit list of things we support like `stopLocationId : ID` and so on. I am a bit unsure here, because I do not know if there is a UI that reads the user input "as is" and passes it to OTP - but I hope not.

optionsome replied:
Yes I wasn't sure if we should include this either in the current form. I'll do some refactoring so we can have coordinate or some entity ID, and additionally a label for that place.

optionsome replied:
I have done something in https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/27e7fb10fa928f560948fed91506eb191d713108 . However, with this model we can't use the oneOf directive without adding another wrapper for the coordinate/stop(/trip) as label can be included with any of those types.

t2gran commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1195,6 +1253,43 @@ input ParkingFilter {
+    place: String
+}
+
+input PlanLocations {
```
Ì would inline these, origin and destination do not share any properties, do they?

t2gran replied:
Unlike many of the other parameters these are both required and important, so I feel it is ok to give them more "space", I would stuff away the less important fields.

t2gran commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -8,10 +8,16 @@ This is only worth it when the execution is long running, i.e. more than ~50 mil
 }
 
+input AccessibilityPreferences {
+    enabled: Boolean
```
If this is the root for all accessibility preferences, then I think there should be a WheelchairPreferences inside it with `transit : { unknown : Cost, inaccessible : Cost }` - not a boolean flagg.

optionsome replied:
Enabling "wheelchair" routing with default parameters isn't simple then if there isn't an on/off switch. We don't yet allow changing the accessibility in other ways through the GraphQL API at least. However, adding those options later is not really a problem as long as we have a wrapper type. I agree that it would make sense to already have a wheelchair field with WheelchairPreferences so we can expand the accessibility later on to cover other things as well.

optionsome replied:
I have added a wrapper type for wheelchair in https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/1a5ae1b705db625888856843ec3725a67f095090

t2gran commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1012,6 +1065,11 @@ type StopGeometries {
 }
 
+input Coordinate {
+    lat: Float!
+    lon: Float!
```
Avoid abbreviations like `lat` and `long`

optionsome replied:
Good point. lat/lon are a bit on the edge as they are pretty standardized abbreviations but using the full words are better anyway.

optionsome replied:
Done in https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/f3a4219870faed100b53be6fddf8acdacf5747d9

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -837,6 +873,23 @@ type Coordinates {
+
+input SearchWindow {
+    dateTime: DateTime
+    timeZone: TimeZone
```
We talking about this at the last meeting and came to the conclusion that we don't want the timezone separately, but rather use a full ISO8601 datetime with offset.

t2gran replied:
I though about it, and for now I think requiring the offset is good. But, it is not so obvious as I first though. There is a long discussion here: https://github.com/opentripplanner/OpenTripPlanner/issues/2602

optionsome replied:
Done in https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/910717a739f12e99b3a445c338fd24a8e4b73637

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1448,6 +1543,11 @@ type Money {
 }
 
+""""
+An ISO-8601-formatted datetime, i.e. `2023-06-13T14:30` for 2:30pm on June 13th 2023.
```
```suggestion
An ISO-8601-formatted datetime with offset, i.e. `2023-06-13T14:30+03:00` for 2:30pm on June 13th 2023 at Helsinki's offset from UTC at that time.
```

leonardehrenfried replied:
We might want to call the scalar `OffsetDateTime` to be super-clear (but we can discuss that).

If we ever need a `LocalDateTime` we can also add that later.

optionsome replied:
Yes I thought too it makes sense to have the offset in the scalar name so at some point we have the datetime name available later on if we need a more generic datetime and the offset in name does make it more obvious that this is a limited part of the datetime spec.

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1836,6 +1936,26 @@ type PageInfo {
+    """CAR"""
+    CAR
+
+    """"Private car trips shared with others."""
```
In our current implementation that is not a street mode. Carpool routes have a schedule and behave like transit.

optionsome replied:
Ok, I'll remove it.

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -2203,6 +2335,41 @@ type Plan {
+type PlanDebugInformation {
+    totalTime: Int
+    pathCalculationTime: Int
+    searchWindowUsed: Long
```
Please use a Duration here.

optionsome replied:
https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/d2f3af5bcb700ec4e04059e676e6a3ccc61b6ab5

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -3694,6 +3893,30 @@ type TicketType implements Node {
     zones: [String!]
 }
 
+scalar TimeZone
```
I don't think we need that anymore, or do we?

optionsome replied:
No, this can be removed.

optionsome replied:
https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/910717a739f12e99b3a445c338fd24a8e4b73637

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -3248,6 +3439,14 @@ enum AbsoluteDirection {
   NORTHWEST
 }
 
+input ScooterPreferences {
```
Are you putting that in just for good measure? I don't think we have that implemented at the moment, isn't it?

optionsome replied:
I added this for two reasons: so it's more clear what settings should be defined when you want to adjust scooter travel and so we can in the future support defining different parameters for cycling/scooter. However, on the downside, what should happen now if both cycling and scooter settings are included.

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -2049,6 +2169,18 @@ type BookingInfo {
+    QUICK
+    SAFE
+    FLAT
+    GREENWAYS
```
`GREENWAYS` is a strange feature that we should probably remove. I have no idea if it does what it says.

optionsome replied:
I think we actually use it in our UI but I haven't really investigated what these different optimization types really did in a long time.

optionsome replied:
We can discuss these optimization types in dev meeting

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -2049,6 +2169,18 @@ type BookingInfo {
+}
+
+enum OptimizationType {
+    QUICK
```
We can debate if we want to rename this to something like "lowest distance" because it can route you, for example, through a road with terrible cobblestones, so you will not be fast.

optionsome replied:
SHORTEST?

leonardehrenfried replied:
i like SHORTEST

leonardehrenfried replied:
but on the other hand, pretty much every other word also has problems

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -2203,6 +2335,41 @@ type Plan {
+
+type PlanDebugInformation {
+    totalTime: Int
+    pathCalculationTime: Int
```
I think we can debate if a Duration is a good idea here. 200ms will be represented as `PT0.2S`. On balance, I think I like it but I can see that not everyone is of the same opinion.

optionsome replied:
Yes I'm not sure about that. Also in more general I wonder if parsing these durations is as easy for clients than just using Int.

optionsome replied:
We can discuss this in dev meeting

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -2203,6 +2335,41 @@ type Plan {
+
+type PlanDateTime {
+    date: DateTime
+    timeZone: TimeZone
```
If we want to add the timezone, we can also think about using Java `ZonedDateTime`'s string format: `2023-06-15T12:43:55+03:00[Europe/Helsinki]`

optionsome replied:
I think we can remove this if we have the offset datetime.

optionsome replied:
https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/910717a739f12e99b3a445c338fd24a8e4b73637

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -2203,6 +2335,41 @@ type Plan {
+}
+
+type PlanDateTime {
+    date: DateTime
```
Should we use an OffsetDateTime?

optionsome replied:
Yes.

optionsome replied:
https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/910717a739f12e99b3a445c338fd24a8e4b73637

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -3694,6 +3893,30 @@ type TicketType implements Node {
+    alight: AlightPreferences
+    transfer: TransferPreferences
+    realtime: RealtimePreferences
+    modes: [TransitModePreference!]
```
We should probably have a think how to integrate Bartosz's filter API here.

optionsome replied:
True. I'm going to check at some point if I'm missing some other latest features from this API in this new draft or any cool features from the transmodel API.

optionsome replied:
I haven't really studied the filter API closely but unless it has major issues, we can port it to this API as is.

leonardehrenfried commented:
I've made lots of comments but I think this is good step in the right direction. Thanks for working on this.

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1448,11 +1538,18 @@ type Money {
+""""
+An ISO-8601-formatted datetime with offset, i.e. `2023-06-13T14:30+03:00` for 2:30pm on June 13th 2023 at Helsinki's offset from UTC at that time.
+"""
+scalar DateTimeWithOffset
```
In java this is called `OffsetDateTime`. Do we want to use this name?

optionsome replied:
I thought about it but OffsetDateTime could be understood to mean a datetime that represents an offset (I know that doesn't make much sense). Although, offset can also mean that something has been "offsetted".

t2gran reviewed (changes requested):
Here are more suggestions from me. To polish this I think creating one or more examples can be very effective.

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -3694,6 +3883,28 @@ type TicketType implements Node {
 }
 
+input TransferPreferences {
+    penalty: Int
+    minimumTime: Duration
```
In the config and in the code we call this transferSlack. Can we use this name?

leonardehrenfried replied:
Or minimumTransferWaitingTime.

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -829,6 +858,13 @@ type Cluster implements Node {
     stops: [Stop!]
 }
 
+input Cost @oneOf {
+    static: Float
+    function: CostFunction
+}
```
I think my thinking here was that having an option to give a static value directly is more understandable than a function that is just a value (although it is a still a valid function). However, I think with documentation we can make it clear that one doesn't have to provide an `x`.

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1012,6 +1065,11 @@ type StopGeometries {
     googleEncoded: [Geometry]
 }
 
+input Coordinate {
```
I thought it was considered bad practice in GraphQL to repeat the "type" of the entry in the name. I havent' checked if it's possible to give the same name to two entries of different types but I can see how that could be slightly confusing.

optionsome replied:
I rechecked this and I think you are right. Having the input suffix seems to be somewhat standard.

optionsome replied:
I somehow managed to mess this up while renaming and used prefix instead of suffix. In the schema we already had some types that used prefix. Should we use suffix or prefix in new input types?

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1448,11 +1538,18 @@ type Money {
+"""
+scalar DateTimeWithOffset
+
 """"
```
The ordering is currently a total mess. I won't be touching it in this pr so the diff is more readable. I'll create a pr afterwards to set a proper order.

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1836,6 +1933,23 @@ type PageInfo {
+    """SCOOTER"""
+    SCOOTER
+
+    """WALK"""
```
I agree, I've copypasted the enum from elsewhere. I'll clean up and improve the comments later.

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -837,6 +873,23 @@ type Coordinates {
     lon: Float
 }
 
+input SearchWindowOptions @oneOf {
+    departure: SearchWindow
+    arrival: SearchWindow
+}
+
+input SearchWindow {
+    dateTime: DateTimeWithOffset
+    duration: Duration
+    numberOfItineraries: Int
+    gracePeriod: SearchWindowGracePeriod
+}
+
+input SearchWindowGracePeriod {
+    duration: Duration
+    numberOfItineraries: Int
+}
+
```
We can discuss this in a developer meeting. I'm really not sure which is better.

optionsome replied:
I've now followed the suggestion  with a minor difference of using the term `dateTime` instead of `time`.

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -3050,6 +3206,24 @@ type QueryType {
+
+input RealtimePreferences {
+    omitCanceled: Boolean
+    ignoreRealtimeUpdates: Boolean
```
I did something like this in https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/8372e112566973d9a99dac64e0645d7251e81464 . Not sure if the naming is better or worse, thought I should not include the REALTIME in the enum values as its already in the name of the enum. I'm not sure if we should still have a wrapper type for the realtime preferences or not. Could we, for example, imagine having some parameter that ignore delays under x minutes for example at some point or some other realtime related parameters that can't really be included in the enum?

optionsome replied:
We discussed this in the last meeting that it could make more sense to enable/disable. Another thing we discussed was planned cancellations. It would be convenient if those could be enabled/disabled from the same field but planned cancellations are not really realtime information. Therefore, should we figure out some other umbrella term for these settings? I drafted a structure like this but the use of the term `Realtime` doesn't fully make sense here:

```
type RealtimePreferences {
    enabled: [RealtimeOption!]
}

enum RealtimeOption {
    REALTIME_ESTIMATES
    REALTIME_CANCELLATIONS
    PLANNED_CANCELLATIONS
}
```

Enabling everything would be done through not defining the parameter.

optionsome replied:
I have now made it an input type with boolean flags again because this enum stuff turned out to be slightly complex and confusing

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -3694,6 +3884,28 @@ type TicketType implements Node {
 }
 
+input TransferPreferences {
+    penalty: Int
```
Please use `cost`.

optionsome replied:
Done

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1836,6 +1933,23 @@ type PageInfo {
 
+enum StreetMode {
+    """BICYCLE"""
+    BICYCLE
```
Add `BICYCLE_RENT` etc...

Don't add qualifier.

leonardehrenfried commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1264,10 +1264,10 @@ input PlanPreferences {
-    access: StreetMode
-    egress: StreetMode
-    transfer: StreetMode
+    direct: [StreetModeInputOptions!]
+    access: [StreetModeInputOptions!]
+    egress: [StreetModeInputOptions!]
+    transfer: [StreetModeInputOptions!]
```
We decided that we want to have a list of items so that we can expand the routingengine later  to use it.

t2gran replied:
Consider: We talked about making `PlanAccessEgressModeInput`, `PlanDirectModeInput` and `PlanTransferModeInput` on todays dev-meeting. That will lead to a lot of duplication and probably also mapping code in the client, but it removes the need for constraint documentation and validation. Not, sure what is best.

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1932,20 +1932,29 @@ type PageInfo {
+
+    CAR_PARKING
+
+    CAR_PICKUP
+
+    CAR_DROPOFF
```
We do

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -3694,6 +3927,28 @@ type TicketType implements Node {
+
+input TransitPreferences {
+    """TODO should we reuse this or make a new design"""
+    unpreferred: InputUnpreferred
```
Did we have support for preferrring certain things already or was it still unimplemented? I'm not sure why I have omitted it from here.

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -758,6 +768,21 @@ enum BikesAllowed {
     NOT_ALLOWED
 }
 
+input BikePreferences {
+    reluctance: Float
+    bikeWalking: WalkPreferences
+    speed: Float
+    switchTime: Duration
+    switchCost: Int
+    optimize: OptimizationCriteria
+    boardCost: Int
+}
```
I should also include these switch parameters for scooter preferences but with what name? I checked the code and it appears these are used when switching to bike walking and back to biking. These are seemingly not used when for example when boarding a rented bike for the first time or parking a bike. We have also had a `bikeWalkingReluctance` parameter. Should we combine at least the cost and reluctance into one cost function parameter. Related note, I'm not convinced we should reuse the `WalkPreferences` for bike/scooter walking as it includes board cost which doesnt make sense in these contexts. Also, can we really set different values for bike walking vs regular walking? There are some things for what it could make sense for like the reluctance or the speed as they typically differ a bit.

t2gran replied:
I agree on the WalkPreferences.

optionsome commented on src/ext/resources/gtfsgraphqlapi/schema.graphqls:
```diff
@@ -1198,6 +1250,43 @@ input ParkingFilter {
+    car: CarPreferences
+    walk: WalkPreferences
+    rental: RentalPreferences
+    parking: VehicleParkingInput
```
Should we reuse this `VehicleParkingInput` or create a new type? The `VehicleParkingInput` had some known limitations but on the other hand, I'm not sure if anyone has interest any time soon to make it more generic.

optionsome commented on src/ext/resources/gtfsgraphqlapi/schema.graphqls:
```diff
@@ -3854,6 +4088,30 @@ type TicketType implements Node {
+input TransferPreferences {
+    costPenalty: Int
+    minimumWaitingTime: Duration
+    maximumAdditionalTransfers: Int = 5
+    maximumTransfers: Int = 12
```
Should we include the maximumTransfers parameter at all as it is slightly problematic and do these defaults make sense (I copied them from the transmodel API)?

optionsome commented on src/ext/resources/gtfsgraphqlapi/schema.graphqls:
```diff
@@ -1213,6 +1306,42 @@ input InputTriangle {
+    Relative importance of cycling friendliness. Factors that are used to evaluate
+    cycling friendliness include safety, convenience and general cyclist preferences.
+    """
+    cyclingFriendliness: WeightingFactor
```
Is friendliness fitting term for "safety"?

leonardehrenfried replied:
we don't need "cycling"

optionsome commented on src/ext/resources/gtfsgraphqlapi/schema.graphqls:
```diff
@@ -3221,6 +3553,19 @@ enum RealtimeState {
+    keepingVehicleAtDestinationCost: Cost
+    allowedNetworks: [String!]
+    bannedNetworks: [String!]
+    pickupTime: Duration
+    pickupCost: Cost
+    dropOffTime: Duration
+    dropOffCost: Cost
```
I wonder if we should expose these or not. Same for the similar parking parameters.

optionsome commented on src/ext/resources/gtfsgraphqlapi/schema.graphqls:
```diff
@@ -4038,6 +4441,38 @@ enum VertexType {
+    or when getting on the vehicle again. However, this is not applied when getting
+    on a rented vehicle for the first time or off the vehicle when returning the vehicle.
+    """
+    hopOnOrOffTimeSeconds: Int
```
I renamed this from switch time to hopOnOrOff and added seconds postfix but alternatively could have used some PositiveWholeSeconds scalar

optionsome commented:
Some things that came to my mind while writing comments:
* Should we call cycling etc speed `maxSpeed` instead of `speed`? It's really common for people to wonder why the average speed of an itinerary doesn't match the configured speed. On the other hand, I haven't looked at the slope algorithm if the speed can be higher than the configured speed.
* Previously we had written some default values in comments but that doesn't really make sense because it depends on the configuration. Should we expose the defaults as its done in the transmodel API so you can fetch all the defaults that are actually in use?
* What does alight slack actually do?
* Can a cost or cost function have negative numbers?
* What format do the locales need to follow?

optionsome commented:
* Should we split PlanAccessEgressMode into different enums for access and egress? Some modes don't seem to make sense in both cases.
* Do we have support for scooter outside of rental?

optionsome commented:
I have created a couple of query examples:
Larger example (this doesn't contain scooter preferences which are similar to bicycle nor all output fields):
<details>
{
  planConnection(
    dateTime: {
      earliestDeparture: "2023-06-13T14:30+03:00"
    }
    searchWindow: {
      duration: "PT2H30M"
      numberOfItineraries: 5
      gracePeriod: {
        duration: "PT10M"
        numberOfItineraries: 1
      }
    }
    origin: {
      location: {
        coordinates: {
          latitude: 10.0
          longitude: 20.0
        }
      }
      label: "Home"
    }
    destination: {
      location: {
        coordinates: {
          latitude: 15.0
          longitude: 25.0
        }
      }
      label: "Work"
    }
    preferences: {
      accessibility: {
        wheelchair: { 
          enabled: true
        }
      }
      street: {
        modes: {
          access: CAR_PARKING
          transfer: WALK
          egress: BICYCLE_RENTAL
          direct: CAR_PARKING
        }
        car: {
          reluctance: 6.5
        }
        bicycle: {
          reluctance: 3.0
          speed: 7.4
          optimize: {
            type: SAFEST_STREETS
          }
          boardCost: 200
          bicycleWalking: {
            speed: 1.3
            reluctance: "100 + 3.5 x"
            hopOnOrOffTimeSeconds: 5
          }
        }
        walk: {
          speed: 2.4
          reluctance: 1.5
          walkSafetyFactor: 0.5
          boardCost: 200
        }
        rental: {
          allowKeepingVehicleAtDestination: true
          keepingVehicleAtDestinationCost: 300
          allowedNetworks: ["foo", "bar"]
        }
        parking: {
          unpreferredCost: 200
          preferred: [{
            select: [{
              tags: ["best-park"]
            }]
          }]
          filters: [{
            not: [{
              tags: ["worst-park"]
            }]
          }]
        }
      }
      transit: {
        unpreferred: {
          routes: ["foo:123"]
          agencies: ["foo:bar"]
          unpreferredCost: "600 + 2.0 x"
        }
        board: {
          waitReluctance: 3.2
          slack: "PT1M30S"
        }
        alight: {
          slack: "PT0S"
        }
        transfer: {
          costPenalty: 200
          minimumTotalSlack: "PT2M"
          maximumAdditionalTransfers: 2
          maximumTransfers: 5
        }
        timetable: {
          includeRealtimeUpdates: true
          includePlannedCancellations: false
          includeRealtimeCancellations: true
        }
        modes: [
          {
            mode: TRAM
            reluctance: 1.3
          },
          {
            mode: BUS
          }
        ]
      }
    }
    locale: "en"
    after: "<some cursor string>"
    first: 3
  ) {
    routingErrors {
      code
      description
      inputField
    }
    pageInfo {
      hasNextPage
      hasPreviousPage
      startCursor
      endCursor
    }
    edges {
      cursor
      node {
        legs {
          startTime
        }
      }
    }
  }
}
</details>

and a simple minimum(?) query:
<details>
{
  planConnection(
    dateTime: {
      earliestDeparture: "2023-06-13T14:30+03:00"
    }
    origin: {
      location: {
        coordinates: {
          latitude: 10.0
          longitude: 20.0
        }
      }
    }
    destination: {
      location: {
        coordinates: {
          latitude: 15.0
          longitude: 25.0
        }
      }
    }
    preferences: {
      street: {
        modes: {
          access: CAR_PARKING
          transfer: WALK
          egress: BICYCLE_RENTAL
          direct: CAR_PARKING
        }
      }
      transit: {
        modes: [
          {
            mode: TRAM
            reluctance: 1.3
          },
          {
            mode: BUS
          }
        ]
      }
    }
  ) {
    edges {
      node {
        legs {
          startTime
        }
      }
    }
  }
}
</details>

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1198,6 +1389,145 @@ input ParkingFilter {
+"""
+Street mode options for different phases of an itinerary.
+"""
+input PlanStreetModesInput {
```
I thought about the `useDefaultTransit ` for a bit. I think there are a few potential issues:
I think it would be better to have the default transit on by default to ease the use of the API
  2. I'm not sure if it explicit enough about how you turn off transit

optionsome commented:
@leonardehrenfried @t2gran I've pushed some commits based on wednesday's meeting and added updated query examples to the summary of this pr.

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1455,6 +1422,28 @@ input PlanPreferencesInput {
+2. As coordinates where the search can start or end but it's not required to use this stop
+    place for a direct transit connnection.
+"""
+input PlanStopLocationInput {
```
Doesn't this need `oneOf`?

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4212,7 +4222,7 @@ enum RealtimeState {
 """
 A cost multiplier for how bad something is compared to being in transit for equal lengths of time.
-Value should be a float and bigger than equal to 1.0.
+The value should be over 0.
```
```suggestion
The value should be greater than 0.
```

optionsome commented:
I tried if it was possible to insert the default values in code. Unfortunately not everything is to modify or copy. I wrote a snippet that is not complete (we need to copy over directives etc) but it does work:
<details>
<pre>
var vehicleParkingInput = typeRegistry.getType("VehicleParkingInput").get();
var updatedInputType = InputObjectTypeDefinition
        .newInputObjectDefinition()
        .name("VehicleParkingInput")
        .inputValueDefinitions(
          vehicleParkingInput
            .getChildren()
            .stream()
            .map(inputValueDefinition -> {
              if (
                "unpreferredCost".equals(((InputValueDefinition) inputValueDefinition).getName())
              ) {
                return ((InputValueDefinition) inputValueDefinition).transform(builder ->
                    builder.defaultValue(IntValue.of(30)).build()
                  );
              }
              return inputValueDefinition;
            })
            .map(InputValueDefinition.class::cast)
            .toList()
        )
        .build();
typeRegistry.remove(vehicleParkingInput);
typeRegistry.add(updatedInputType);
</pre>
</details>

Some of this nasty stuff can be hidden behind some utility method. Still, we would be doing something that is not really standard and it would increase the complexity. Is this a trade-off we are willing to accept in order to get the actual defaults into the schema?

optionsome commented:
The major topics that still need decisions:
* What to do with rental and parking preferences?
   * Should they all be put inside street mode specific preferences so they can be set different per mode? Should some parameters (network/tag filtering) be done per "request"?
* Where/how should we define modes? Is flex a street or transit mode or something else?
* Should we make it possible to define multiple requests/queries within same plan query with different sets of modes and some(?) preferences that go through the same itinerary filtering process?
* Do we include transit unpreferred in the first version of this new plan endpoint in the current format or in any format?

optionsome commented:
We decided yesterday that we will put all the logic for inserting the default values into the plan arguments in code instead of using some directive in schema for it.

t2gran commented:
```
plan
  modes : PlanModesInput
    directOnly = false
    transitOnly = false
    direct : DirectMode[]
    transit : TransitModesInput
      access : AccessMode[]
      transit : TransitMode[]
      transfer : TransferMode[]
      egress : EgressMode[]
```
**Examples**
```
plan()                             =>  default direct and transit
plan(modes: {})                    =>  default direct and transit
plan(modes: {directOnly:true})     =>  default direct
plan(modes: {transitOnly:true})    =>  default transit
plan(modes: {direct: X})           =>  direct: X and default transit
plan(modes: {transit: {access:X})  =>  default direct, default transit, except access=X

```
There is not good way around the flags. But, I think setting the flags is less common - the most common use-case is to search for both direct and transit. Also, this has the advantage that there is no ambiguity or magic values. Having two flags also make this symmetrical - no supprises.

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1233,6 +1452,184 @@ input ParkingFilter {
+    """
+    itineraryFilterDebugProfile: ItineraryFilterDebugProfile = OFF
+
+    """
+    Pick one itinerary from each group after putting itineraries that are `85%` similar together,
+    if the given value is `0.85`, for example. Itineraries are grouped together based on relative
+    the distance of travel that is identical between the itineraries.
+    """
+    groupSimilarityKeepOne: Float = 0.85
+
+    """
+    Pick three itineraries from each group after putting itineraries that are `68%` similar together,
+    if the given value is `0.68`, for example. The itineraries are grouped by similar legs
+    (on board same journey). The value must be at least `0.5`.
+    """
+    groupSimilarityKeepThree: Float = 0.68
```
Should we use the WeightingFactor scalar for these, create a new one or just use Float? I copied the comments and slightly modified them from the transmodel API but there are some aspects I don't understand:
1. Are they really grouped together based on the distance of travel they share, not duration? Do all legs count or only transit legs?
2. Does `groupSimilarityKeepOne` also have a lower limit?

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1233,6 +1452,184 @@ input ParkingFilter {
+Settings that control the behavior of itinerary filtering. These settings are meant mainly
+for debugging purposes.
+"""
+input PlanItineraryFilterSettingsInput {
```
I copied some fields from the transmodel API for this but I didn't copy `transitGeneralizedCostLimit` as I didn't understand it's documentation completely and not sure if we should actually have a DoubleFunction scalar for it or also somehow split that into an input type with two fields?

optionsome commented:
This pr is now updated to follow all the decisions made on wednesday (which are documented in https://github.com/opentripplanner/OpenTripPlanner/pull/5185#issuecomment-1805261884) and the mode input model created by @t2gran . I have updated the examples in the summary.

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -2382,6 +3275,69 @@ type Plan {
+    """
+    searchDateTime: OffsetDateTime
+
+    """
+    Origin of the itinerary search.
+    TODO figure out the best output type
+    """
+    origin: Place!
+
+    """
+    Destination of the itinerary search.
+    TODO figure out the best output type
+    """
+    destination: Place!
```
The issue with these places is that `Place` type should not be used since it has mandatory arrival and departure time fields. We can't use `PlaceInterface` either because it's used in the nearest places search and in the future, we want to add trip (or a stop at a trip) as a location for the plan query and the trip doesn't make sense as a potential output for the nearest search and we also need a coordinate location type. Also, we potentially could want to return both the original name of a stop for example and the label user has given for the stop (the Stop type itself doesn't have both).

What we could do is add another interface/union that is only used in this context. However, that still doesn't solve the issue that we want both the name of the location and the user given label for it. Therefore, I suggest the following schema:
```
origin: PlanLabeledLocation!

destination: PlanLabeledLocation!
```
```
type PlanLabeledLocation {  
    label: String
    
    // This can be null if for example user provides a stop id but that stop doesn't exist.
    location: PlanLocation
}
```

```
union PlanLocation =  Stop | PlanCoordinates
```

```
type PlanCoordinates {
    latitude: Coordinate!

    longitude: Coordinate!
}
```

optionsome replied:
* The existing Coordinates type has the issue that it doesn't use the Coordinate scalar. I was a bit struggling with the naming conventions because I wanted the scalar to be called Coordinate but I think CoordinateValue is a good idea. I think I've gone back and worth with naming the coordinate `coordinate` or `coordinates` but I agree with you that it should be called `coordinate`. I think I ended up in the current naming by accident.
* I was worried the PlanLocation would be too generic as we might end up having a location without label at some point for some purpose but I think the locations within itineraries should be called ItineraryLocation or whatever so it's not as risky. I'm not sure what is the best practice with naming unions but having the "type" name as a postfix feels a bit wrong but it seems to be a convention with the input at least.

testower replied:
This is a bit contrary to common language use. A coordinate (singular) is usually understood as the x or y part, and the tuple (x,y) is called coordinates (plural). So you have the latitude *coordinate* and the longitude *coordinate* and together they form the *coordinates*. 

https://www.oxfordlearnersdictionaries.com/definition/american_english/coordinate_2

optionsome replied:
I was under the same impression as @t2gran that coordinate would be the correct term for a lat lon pair even though coordinates is used for it often but now when I try to look information about it, I can mostly just find information pointing to the other direction.

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1233,6 +1452,222 @@ input ParkingFilter {
+    """
+    Coordinates of the location. Note, either coordinates or a stop location should be defined.
+    """
+    coordinates: PlanCoordinatesInput
```
```suggestion
    coordinate: PlanCoordinateInput
```

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1896,38 +2411,108 @@ enum Mode {
 
-    """COACH"""
+    """
+    Sometimes coaches use this mode but sometimes they are modelled to use `BUS` mode
+    instead.
```
Honestly, I don't think I have seen a data set where they would use this COACH type.

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1896,38 +2411,108 @@ enum Mode {
-    """RAIL"""
+    """
+    This includes long or short distance trains. Tram, subway, monorail and funicular have
+    their own modes but it's possible that sometimes they use this mode instead.
```
I'm ok with deleting this. I'm not even sure where I invented this from.

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -2382,6 +3312,67 @@ type Plan {
+    Errors faced during the routing search.
+    """
+    routingErrors: [RoutingError!]!
+
+    """
+    The search window that was used for the search.
+    """
+    searchWindowUsed: Duration
```
The PageInfo is a standard relay pagination thing, I'm not sure if we can/should expand it to include other things..

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4153,6 +5487,121 @@ enum VertexType {
+    bicycle walking. However, this cost is not applied when getting on a rented bicycle
+    for the first time or when getting off the bicycle when returning the bicycle.
+    """
+    hopOnOrOffCost: Cost
```
```suggestion
    hopCost: Cost
```

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4153,6 +5487,121 @@ enum VertexType {
+    scooter walking. However, this cost is not applied when getting on a rented scooter
+    for the first time or when getting off the scooter when returning the scooter.
+    """
+    hopOnOrOffCost: Cost
```
```suggestion
    hopCost: Cost
```

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -3492,6 +4648,44 @@ enum OccupancyStatus {
+    What criteria should be used when optimizing a scooter route.
+    """
+    optimization: ScooterOptimizationInput
+
+    """
+    Cost of boarding a vehicle with a scooter.
+    """
+    boardCost: Cost
```
```suggestion
```

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1983,7 +2537,266 @@ type PageInfo {
+    can be defined in GTFS-Flex or in Netex. Note, this can include walking before
+    or after the flexible transit leg.
+    """
+    FLEX
+
+    """
+    Riding a scooter from the origin to the destination. Note, this can include walking
+    when it's needed to walk the scooter.
+    """
+    SCOOTER
```
```suggestion
    FLEX
```

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1983,7 +2537,266 @@ type PageInfo {
+    can be defined in GTFS-Flex or in Netex. Note, this can include walking before
+    or after the flexible transit leg.
+    """
+    FLEX
+
+    """
+    Riding a scooter to a stop and boarding a vehicle with the scooter.
+    Note, this can include walking when it's needed to walk the scooter. Access
+    can use scooter only if the mode used for transfers and egress is also `SCOOTER`.
+    """
+    SCOOTER
```
```suggestion
    FLEX
```

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1983,7 +2537,266 @@ type PageInfo {
+    can be defined in GTFS-Flex or in Netex. Note, this can include walking before
+    or after the flexible transit leg.
+    """
+    FLEX
+
+    """
+    Riding a scooter from a stop to the destination. Note, this can include walking
+    when it's needed to walk the scooter. Egress can use scooter only if the mode
+    used for transfers and egress is also `SCOOTER`.
+    """
+    SCOOTER
```
```suggestion
    FLEX
```

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1983,7 +2537,266 @@ type PageInfo {
+    include walking when it's needed to walk the bicycle. Transfers can only use
+    cycling if the mode used for access and egress is also `BICYCLE`.
+    """
+    BICYCLE
+
+    """
+    Riding a scooter between transit vehicles (typically between stops). Note, this can
+    include walking when it's needed to walk the scooter. Transfers can only use
+    scooter if the mode used for access and egress is also `SCOOTER`.
+    """
+    SCOOTER
```
```suggestion
    BICYCLE
```

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -333,6 +354,63 @@ type BikePark implements Node & PlaceInterface {
+"""
+input DestinationBicyclePolicyInput {
+    """
+    Is it possible to arrive to the destination without dropping off the rental
+    bicycle first.
```
```suggestion
    For networks that require station drop-off, should the routing engine offer results that go directly to the destination without dropping off the rental bicycle first.
```

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -333,6 +354,63 @@ type BikePark implements Node & PlaceInterface {
+"""
+input DestinationScooterPolicyInput {
+    """
+    Is it possible to arrive to the destination without dropping off the rental
+    scooter first.
```
```suggestion
    For networks that require station drop-off, should the routing engine offer results that go directly to the destination without dropping off the rental scooter first.
```

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -863,6 +1048,27 @@ type Coordinates {
+    Latest arrival time date time. The returned itineraries should not
+    arrive to the destination after this instant unless one is using
+    paging to find later itineraries. Note, it is not currently possible
+    to define both  `earliestDeparture` and `latestArrival`.
```
```suggestion
    to define both `earliestDeparture` and `latestArrival`.
```

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1173,6 +1393,59 @@ input InputPreferred {
 }
 
+"""
+Locale in string format. For example, `en` or `en-US`.
```
```suggestion
Locale in the format defined in [RFC5646](https://datatracker.ietf.org/doc/html/rfc5646). For example, `en` or `en-US`.
```

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1233,6 +1506,243 @@ input ParkingFilter {
 }
 
+"""
+Settings that control the behavior of itinerary filtering.  These are advanced settings and
+should not be set by a user through user preferences.
```
```suggestion
Settings that control the behavior of itinerary filtering. **These are advanced settings and
should not be set by a user through user preferences.**
```

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1233,6 +1506,243 @@ input ParkingFilter {
+    """
+    itineraryFilterDebugProfile: ItineraryFilterDebugProfile = OFF
+
+    """
+    Pick one itinerary from each group after putting itineraries that are `85%` similar together,
+    if the given value is `0.85`, for example. Itineraries are grouped together based on relative
+    the distance of transit travel that is identical between the itineraries (access, egress and
+    transfers are ignored). The value must be at least `0.5`.
+    """
+    groupSimilarityKeepOne: Ratio = 0.85
+
+    """
+    Pick three itineraries from each group after putting itineraries that are `68%` similar together,
+    if the given value is `0.68`, for example. Itineraries are grouped together based on relative
+    the distance of transit travel that is identical between the itineraries (access, egress and
+    transfers are ignored). The value must be at least `0.5`.
+    """
+    groupSimilarityKeepThree: Ratio = 0.68
+
+    """
+    Of the itineraries grouped to maximum of three itineraries, how much worse can the non-grouped
+    legs be compared to the lowest cost. `2.0` means that they can be double the cost, and any
+    itineraries having a higher cost will be filtered away. Use a value lower than `1.0` to turn the
+    grouping off.
+    """
+    groupedOtherThanSameLegsMaxCostMultiplier: Float = 2.0
```
Do we actually want to expose them? What are they used for?

t2gran replied:
There are two reasons for exposing these:
 - If you want to experiment with them.
 - These are probably the most important filter settings for transit. The end user should not be exposed, but a you might want to set these in a BFF. Note the doc on itinerary filter settings: 
 > Configure the itinerary-filter-chain. NOTE! THESE PARAMETERS ARE USED FOR SERVER-SIDE TUNING AND IS AVAILABLE HERE FOR TESTING ONLY.

leonardehrenfried reviewed (approved):
I read through this PR again and I'm satisfied that we have found a good new schema for specifiying the plan request.

I've left a few suggestions for the documentation.

optionsome commented:
@leonardehrenfried `@magidoc/cli` complains about the use of `@oneOf` directive. What should we do about it? I already tried using the latest version of it and the problem still persists. Java graphql library has support for the directive already. I wonder if it is a problem if we still define the directive in the schema?

optionsome commented:
I looked at the docs, maybe a solution would be to run the OTP server in the actions and use the URL instead of the static schema file?

leonardehrenfried commented:
If magidocs doesn't support it I would rather switch tools than to dumb down our schema. However, do you know if oneOf has been merged into the spec already? https://github.com/graphql/graphql-spec/pull/825

optionsome commented:
Seems like nothing at least explodes if I add the directive to the schema file. I think I'll do it since I guess the directive has not been merged to the spec yet.

I also re-read the spec again. This part made me wonder if we should or should not use the directive on the bicycle/scooter optimization types as we want to have a default value:

```
5. If the Input Object is a OneOf Input Object then:
      1. The type of the input field must be nullable.
      2. The input field must not have a default value.
```

leonardehrenfried commented:
I've created a ticket over at https://github.com/magidoc-org/magidoc/issues/319

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -3260,6 +4225,115 @@ type QueryType {
+        is a possibility that more itineraries are used later on (limiting the number of returned itineraries
+        does not affect the performance by a lot but can affect the network bandwidth usage).
+        """
+        numberOfItineraries: Int
```
What did we decide about this being required/having a default value? I can't remember anymore if we made some decision or not.

optionsome replied:
or did we decide to not limit it by default at all and just limit through the search window?

t2gran replied:
The `numberOfItineraries` should have a default value (set in code). It is used to crop the searchwindow AFTER the search is done so setting it too low will result in poor performance. If a client can only show 3 results then 3 is ok, if many results are allowed then setting it to 20/50 is fine. Note! this only reduces the amount of data retuned - it does not limit the search.

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1253,6 +1528,243 @@ input ParkingFilter {
+    transfer: [PlanTransferMode!]
+
+    """
+    Transit modes and reluctancies associated with them. Each defined mode can be used in
```
```suggestion
    Transit modes and reluctances associated with them. Each defined mode can be used in
```

optionsome commented:
This is now ready for review again. Tests are still missing and the OffsetDateTime implementation is not 100% up-to-date with what @leonardehrenfried has been developing in https://github.com/opentripplanner/OpenTripPlanner/pull/5660. I'll wait until it's merged to update this.

There are a couple of things I'll plan to implement in separate prs:
1. Injecting defaults into the schema
2. Support for defining multiple street modes for direct, access, egress and transfer

optionsome commented on src/main/java/org/opentripplanner/apis/gtfs/mapping/RouteRequestMapper.java:
This file is completely rewritten. The old code is in LegacyRouteRequestMapper but git got confused. When reviewing this, it's better to review the file as whole instead of looking at the changes.

leonardehrenfried commented on src/main/java/org/opentripplanner/apis/gtfs/GraphQLScalars.java:
```diff
@@ -152,4 +336,138 @@ public Grams parseLiteral(Object input) throws CoercingParseLiteralException {
+        public Double serialize(@Nonnull Object dataFetcherResult)
+          throws CoercingSerializeException {
+          var validationException = new CoercingSerializeException(
+            "Value is under 0 or greater than 1."
```
```suggestion
            "Value is less than 0 or greater than 1."
```

leonardehrenfried commented on src/main/java/org/opentripplanner/apis/gtfs/GraphQLScalars.java:
```diff
@@ -152,4 +336,138 @@ public Grams parseLiteral(Object input) throws CoercingParseLiteralException {
+          if (input instanceof Double doubleValue) {
+            return validateRatio(doubleValue)
+              .orElseThrow(() ->
+                new CoercingParseValueException("Value is under 0 or greater than 1.")
```
```suggestion
                new CoercingParseValueException("Value is less than 0 or greater than 1.")
```

leonardehrenfried commented on src/main/java/org/opentripplanner/apis/gtfs/mapping/DirectModeMapper.java:
```diff
@@ -0,0 +1,24 @@
+      case BICYCLE_PARKING -> StreetMode.BIKE_TO_PARK;
+      case CAR -> StreetMode.CAR;
+      case CAR_RENTAL -> StreetMode.CAR_RENTAL;
+      case CAR_PARKING -> StreetMode.CAR_TO_PARK;
```
Does a direct mode of CAR_PARKING mean that you want to look for a parking lot near the destination and walk the rest?

optionsome replied:
I think so yes. It's up to debate if we should support the functionality but pretty sure we do.

leonardehrenfried commented on src/main/java/org/opentripplanner/apis/gtfs/mapping/RouteRequestMapper.java:
```diff
@@ -40,227 +43,51 @@ public static RouteRequest toRouteRequest(
-          filterRequestBuilder.addNot(SelectRequest.of().withAgenciesFromString((String) s).build())
+    var args = new GraphQLTypes.GraphQLQueryTypePlanConnectionArgs(environment.getArguments());
+    var dateTime = args.getGraphQLDateTime();
+    if (dateTime.getGraphQLEarliestDeparture() != null) {
```
Does the framework validate that you cannot set both latestDeparture and earliestArrival?

leonardehrenfried commented on src/main/java/org/opentripplanner/apis/gtfs/mapping/RouteRequestMapper.java:
```diff
@@ -283,132 +110,616 @@ private static Stream<String> parseOperation(Collection<Map<String, Collection<S
+   * getting a list of input objects is not possible through using the generated types in
+   * {@link GraphQLTypes}.
+   * <p>
+   * TODO this ugliness can be removed when the bug gets fixed
```
Is there a link to a ticket?

optionsome replied:
I haven't created a ticket yet at least yet. Haven't really looked into this issue recently, so no clue how easily it could be fixed.

leonardehrenfried commented on src/main/java/org/opentripplanner/apis/gtfs/mapping/RouteRequestMapper.java:
```diff
@@ -283,132 +110,616 @@ private static Stream<String> parseOperation(Collection<Map<String, Collection<S
+  ) {
+    if (args != null) {
+      var allowedNetworks = args.getGraphQLAllowedNetworks();
+      if (allowedNetworks != null && allowedNetworks.size() > 0) {
```
This check happens over and over. Can you extract something like `CollectionUtils.hasValue`?

optionsome replied:
Empty array is returned (which makes sense). I've used the utils now.

leonardehrenfried commented on src/main/java/org/opentripplanner/apis/gtfs/mapping/RouteRequestMapper.java:
```diff
@@ -283,132 +110,616 @@ private static Stream<String> parseOperation(Collection<Map<String, Collection<S
+  }
+
+  /**
+   * TODO this doesn't support multiple street modes yet
```
Can you throw an exception when there are multiple?

optionsome replied:
Not sure if we want that. For example, we could define `[BICYCLE_RENTAL, WALK]` already now and it would work now and after we add support. However, if you have the modes in different order, it wouldn't use bike rental now. We can discuss this in a dev meeting or here.

leonardehrenfried replied:
If you use BICYCLE_RENTAL then WALK is automatically used as a fallback and [BICYCLE_RENTAL,WALK] is the only combination that we support right now.

optionsome replied:
Yep, therefore if I would now throw an exception if more modes are defined, client code would need to be changed if we filter out non bike rental legs if just BICYCLE_RENTAL is defined. However, as @t2gran pointed out in the dev meeting, we can do some "breaking" changes while experimenting.

leonardehrenfried commented on src/main/java/org/opentripplanner/apis/gtfs/mapping/RouteRequestMapper.java:
```diff
@@ -283,132 +110,616 @@ private static Stream<String> parseOperation(Collection<Map<String, Collection<S
+      }
+
+      var transfer = transit.getGraphQLTransfer();
+      if (transfer != null && transfer.size() > 0) {
```
Again, lots of identical checks. Please extract a utility method.

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -863,6 +1032,27 @@ type Coordinates {
+    """
+    Earliest departure date time. The returned itineraries should not
+    depart before this instant unless one is using paging to find earlier
+    itineraries. Note, it is not currently possible to define both
```
Currently? Is it ever going to be supported?

optionsome replied:
Well in theory we could. We quickly discussed all sorts of weird date time/number of itineraries limitations (partly inspired by the OJP specification).

leonardehrenfried commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -863,6 +1032,27 @@ type Coordinates {
+    """
+    Latest arrival time date time. The returned itineraries should not
+    arrive to the destination after this instant unless one is using
+    paging to find later itineraries. Note, it is not currently possible
```
same here.

leonardehrenfried reviewed (changes requested):
This is getting close - thanks for not giving up!

I have a few requests and then we have to decide how to deal with multiple access/egress/transfer modes.

leonardehrenfried commented:
However, to get a feeling for the API, can you add one of those GraphQL test?

leonardehrenfried commented:
Also, while I don't expect a test for all properties in the mapper, I would like to see a few that test the basic functionality. The rest can come when we discover the bugs. :)

t2gran commented:
You should add tests for:
 - Mappers
 - Utils 
 - Scalars
 
Module and integration tests are optional. If you would like to merge this so you can continue on testing it with clients, then you should create a issue/follow up PR with the tests - this is considered work-in-progress, and it is up to you which order you do stuff as long as it is done in the end.

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -3981,6 +5082,115 @@ type TicketType implements Node {
+    realistic as the transfers could be invalid due to delays,
+    trips can be cancelled or stops can be skipped.
+    """
+    excludeRealTimeUpdates: Boolean = false
+
+    """
+    When true, departures that have been cancelled ahead of time will be
+    included during the routing. This means that an itinerary can include
+    a cancelled departure while some other alternative that contains no cancellations
+    could be filtered out as the alternative containing a cancellation would normally
+    be better.
+    """
+    includePlannedCancellations: Boolean = false
+
+    """
+    When true, departures that have been cancelled through a real-time feed will be
+    included during the routing. This means that an itinerary can include
+    a cancelled departure while some other alternative that contains no cancellations
+    could be filtered out as the alternative containing a cancellation would normally
+    be better. This option can't be set to true while `includeRealTimeUpdates` is false.
+    """
+    includeRealTimeCancellations: Boolean = false
```
Should these have hard-coded default values in the schema? Currently Transmodel API has them. It's possible to configure value for `excludeRealtimeUpdates` (`ignoreRealtimeUpdates`) in default route request.

optionsome replied:
We decided to remove the defaults. I have done the same for number of itineraries as it can be configured as well.

optionsome commented on src/main/java/org/opentripplanner/apis/gtfs/GraphQLScalars.java:
```diff
@@ -152,4 +336,138 @@ public Grams parseLiteral(Object input) throws CoercingParseLiteralException {
+        @Override
+        public Double parseValue(Object input) throws CoercingParseValueException {
+          if (input instanceof Double doubleValue) {
+            if (Double.doubleToRawLongBits(doubleValue) < 0) {
```
What should be the minimum value for reluctance? Does letting it be 0 ever make sense?

optionsome replied:
We decided it should be 0.1 and I have changed the value.

leonardehrenfried commented:
Please resolve the conflicts.

optionsome commented:
@t2gran regarding the direct mode discussion from dev meeting, I'll list the potential cases below:
- User wants direct suggestions
   - This is done with the current schema by either not defining anything (walk is then returned) or by defining some direct mode under `modes`
- User doesn't want direct suggestions
   - This is done by setting `transitOnly` as true under `modes`
- User wants only direct suggestions
   - This is done by setting `directOnly` as true under `modes`
- User wants to only use direct itineraries for filtering but not return them
   - This can't be done currently. Perhaps suitable places for relevant parameters would be under `modes` or `itineraryFilter`
- User wants direct itineraries but doesn't want transit itineraries be filtered out due to comparison to direct
  - This can't be done currently. Perhaps suitable places for relevant parameters would be under `modes` or `itineraryFilter`

optionsome commented:
I have now added all the tests I had planned to add and validation for the street modes. I'm unsure if/how I should update the street mode documentation in the schema about these current limitations.

I will fix the conflict later, when this pr is ready to merged as there might be a proper release for the dependency at that point. There is still the discussion from https://github.com/opentripplanner/OpenTripPlanner/pull/5185#issuecomment-2012675499 which needs to be resolved.

optionsome commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1253,6 +1510,213 @@ input ParkingFilter {
+    an itinerary but doesn't have to be. If direct search is not disabled, there can be an
+    itinerary without any transit legs. By default, all transit modes are usable.
+    """
+    transit: [PlanTransitModePreferenceInput!]
```
Btw what should an empty list mean? Right now it's just ignored and default modes are used. We have explicit field for disabling transit search. If the list is empty, we could:
1. Throw an exception
2. Disable transit routing
3. Use default mode
This isn't unique to this field. We should make some decision on how empty lists should be interpreted in general. I think I might have ignored the input in all cases but maybe I shouldn't have. For example, when banning some things there isn't other way to override default banning.

leonardehrenfried replied:
Until we have worked out what it means we should throw an exception so that whatever expectation that people have doesn't get cemented.

Can you remind me why we went with an explicit field for enabling/disabling transit?

optionsome replied:
@t2gran came up with this design but I think the reason is that it's more explicit.

leonardehrenfried commented:
I'm satisfied with the schema, implementation and tests and I'm happy to approve and work out the rest after a bit of testing.

leonardehrenfried commented:
Can you resolve the conflicts?

optionsome commented:
This updates graphql-java which contains a lot of changes so someone should check that the micrometer integration still works as intended (and the transmodel API).

t2gran commented on src/main/java/org/opentripplanner/apis/gtfs/mapping/routerequest/BicyclePreferencesMapper.java:
```diff
@@ -0,0 +1,157 @@
+    GraphQLTypes.GraphQLBicyclePreferencesInput args,
+    DataFetchingEnvironment environment
+  ) {
+    if (args != null) {
```
Code style, I like precondition not long if-blocks without an else:
```
if(args == null) { return; }

// main logic
```

optionsome replied:
I've refactored the ones I quickly found in https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/f8dc495a83a89ba628e2a899330615dc589405e2 but let me know if you find more of these.

t2gran commented on src/test/java/org/opentripplanner/apis/gtfs/mapping/routerequest/RouteRequestMapperScooterTest.java:
```diff
@@ -0,0 +1,204 @@
+    var scooterTime = 0.2;
+    scooterArgs.put(
+      "preferences",
+      Map.ofEntries(
+        entry(
+          "street",
+          Map.ofEntries(
+            entry(
+              "scooter",
+              Map.ofEntries(
+                entry(
+                  "optimization",
+                  Map.ofEntries(
+                    entry(
+                      "triangle",
+                      Map.ofEntries(
+                        entry("safety", scooterSafety),
+                        entry("flatness", scooterFlatness),
+                        entry("time", scooterTime)
+                      )
+                    )
+                  )
+                )
+              )
+            )
+          )
+        )
+      )
```
The utility methods and test is in the same class because it was easy for me to do, and post - I would splitt it. If you put the utility class in the test code or main does not matter much.

optionsome replied:
Otherwise I like this idea, but the problem is that we can't really parse into custom objects such as OffsetDateTime from a string in the json as the `toMap` is not aware of the schema/scalars. Perhaps it would be possible to use the graphql code to parse json into a map or directly into a `DataFetchingEnvironment`. Do you want me to investigate that possibility?

t2gran replied:
Ok, not know. But, we should find a solution for this. But, we could take one step back and look at the best way to automate and test the API.

optionsome commented on src/main/java/org/opentripplanner/apis/gtfs/model/PlanPageInfo.java:
```diff
@@ -0,0 +1,12 @@
+
+public record PlanPageInfo(
+  ConnectionCursor startCursor,
+  ConnectionCursor endCursor,
```
https://github.com/opentripplanner/OpenTripPlanner/pull/5185/commits/22a976301224299c0e1061775b6ec12ff78d192b is this now correct?

t2gran commented on src/main/java/org/opentripplanner/apis/gtfs/model/PlanPageInfo.java:
```diff
@@ -0,0 +1,77 @@
+    boolean hasNextPage,
+    Duration searchWindowUsed
+  ) {
+    this.startCursor = startCursor;
+    this.endCursor = endCursor;
```
```suggestion
    this.startCursor = Objects.requireNonNull(startCursor);
    this.endCursor = Objects.requireNonNull(endCursor);
```
Prevent failing in eq/hc by failing fast in constructor.

optionsome replied:
The cursors are nullable and we shouldn't prevent building this object if they are null. However, thanks for pointing this out as the eq/hc were not null safe, I've updated them now.

t2gran replied:
Had a feeling these where nullable, but the code told another story - happy to catch the bug.

t2gran commented on src/main/java/org/opentripplanner/apis/gtfs/model/PlanPageInfo.java:
```diff
@@ -0,0 +1,91 @@
+  @Override
+  public int hashCode() {
+    return Objects.hash(
+      startCursor != null ? startCursor.getValue() : null,
+      endCursor != null ? endCursor.getValue() : null,
```
```suggestion
      startCursor == null ? null : startCursor.getValue(),
      endCursor == null ? null : endCursor.getValue(),
```
This is nitpicking, but the above change is easier to read and avoid unnecessary negation.

t2gran commented on src/main/java/org/opentripplanner/apis/gtfs/model/PlanPageInfo.java:
```diff
@@ -0,0 +1,91 @@
+      Objects.equals(searchWindowUsed, that.searchWindowUsed) &&
+      hasPreviousPage == that.hasPreviousPage &&
+      hasNextPage == that.hasNextPage
+    );
```
The above code is not readable. I know IntelliJ probably suggest to "merge" if´s. In this case you also repeat the logic for the `start` and `end` cursor - a method like: `private static boolean equalsCursors(ConnectionCursor a, ConnectionCursor b)` could help - also you get a place to add a comment on why you are not calling the `ConnectionCursor.equals()`.

t2gran reviewed (changes requested):
This looks good, my last comments are nit-picking and I am happy to approve without fixing them - let me know what you prefere.

t2gran commented on src/main/java/org/opentripplanner/apis/gtfs/model/PlanPageInfo.java:
```diff
@@ -0,0 +1,91 @@
+      return false;
+    }
+    PlanPageInfo that = (PlanPageInfo) o;
+    return (
+      (
+        (startCursor == null && that.startCursor == null) ||
+        (
+          startCursor != null &&
+          that.startCursor != null &&
+          Objects.equals(startCursor.getValue(), that.startCursor.getValue())
+        )
+      ) &&
+      (
+        (endCursor == null && that.endCursor == null) ||
+        (
+          endCursor != null &&
+          that.endCursor != null &&
+          Objects.equals(endCursor.getValue(), that.endCursor.getValue())
+        )
+      ) &&
+      Objects.equals(searchWindowUsed, that.searchWindowUsed) &&
+      hasPreviousPage == that.hasPreviousPage &&
+      hasNextPage == that.hasNextPage
+    );
```
The above code is not readable. I know IntelliJ probably suggest to "merge" if´s. In this case you also repeat the logic for the start and end cursor - a method like: private static boolean equalsCursors(ConnectionCursor a, ConnectionCursor b) could help - also you get a place to add a comment on why you are not calling the ConnectionCursor.equals().

t2gran replied:
PS! This was originally part of the the above review, but I resubmitted it with an update...

t2gran commented on src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -1268,6 +1525,213 @@ input ParkingFilter {
+    """
+    Should only the direct search without any transit be done.
+    """
+    directOnly: Boolean = false
+
+    """
+    Should only the transit search be done and never suggest itineraries that don't
+    contain any transit legs.
+    """
+    transitOnly: Boolean = false
```
We should probably have split the transit search(T) and street search(S) in to queries. 
 - T is time dependent and has a search-window, S is most of the time time-shiftable 
 - The S result is always first
 - The S is excluded when paging
 - The only reason to have it together is to use the S to prune the T - but we could probably find other ways to do that.

optionsome replied:
Do you mean two graphql queries or queries within OTP? Rerunning the street routing is indeed unnecessary when paging. We have thought about some caching solutions for this previously.

Is this comment related to @leonardehrenfried 's question about where the penalty should belong to?

leonardehrenfried reviewed (approved):
The one remaining question I had was where `accessEgressPenalty` should live. After a bit of deliberation, we decided in today's meeting that we will leave the structure as it is and find a place should we need to add the penalties.

optionsome commented:
Thanks for spending time on reviewing this. We need to decide when we deprecate the old query and remove the deprecation from this one but I think we can wait some weeks at least.

arnoclr commented:
Hello, I can't find any mention of this, but where are the old options of `plan` to ban certain stops or routes in this new `planConnection` ? https://docs.opentripplanner.org/api/dev-2.x/graphql-gtfs/types/InputBanned

optionsome commented:
> Hello, I can't find any mention of this, but where are the old options of `plan` to ban certain stops or routes in this new `planConnection` ? https://docs.opentripplanner.org/api/dev-2.x/graphql-gtfs/types/InputBanned

If I remember correctly, it isn't implemented yet because we wanted to limit the scope of this pr and we didn't necessarily want to implement it in the same way as it was implemented before. Instead, we want to have some sort of a bit more sophisticated filter structure. You are the second person to ask for this feature in the last couple of weeks. I'll bring this topic up in a [developer meeting](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/CONTRIBUTING.md#developer-meetings) within a week or two. You can also show up in a meeting if you want and we can discuss this.
